### PR TITLE
Cache observance transitions in Timezone.get_observance

### PR DESCRIPTION
**Pull request by FuNK3Y**
## Description

`Timezone.get_observance()` re-expands the observance recurrence **from the start on every call**. For Microsoft Exchange / Office 365 "Customized Time Zone" `VTIMEZONE`s, the observance `DTSTART` is in year **1601** with an unbounded `YEARLY` rule, so each offset lookup walks ~400+ years of yearly transitions. A calendar with many events triggers this per event.

Importing a simple Outlook.com calendar in home assistant using the remote calendar blew away my whole Home Assistant. I asked Claude for the fix, and this one looks pretty persuasive and reduced the processing time significantly (17s => 0.005s)

## Root cause

`get_observance` calls `self._observances()` — which builds fresh `RecurIterable(observance.as_ruleset())` iterators and expands them from `DTSTART` (1601) — and walks it linearly to the queried value, **with no caching**, on every call.

This is distinct from the path fixed in #592: that cached `TzInfo.dst()` for timezones that **map to a named `TzInfo`** (e.g. `Pacific Standard Time`). `"Customized Time Zone"` has no canonical name, so it stays an embedded timezone resolved through `Timezone.get_observance` and was not covered.

## Fix

Cache the observance transitions (sorted ascending), extending the cache lazily only as far as the highest datetime queried, then `bisect` it. The recurrence is expanded once instead of per call; repeated lookups become `O(log n)`.

## Results

Measured against the real 81-event Outlook calendar (`ical` parse + `materialize_timeline`):

| | before | after |
| --- | --- | --- |
| `materialize_timeline` | 17.6s | 0.005s |

Outputs are byte-for-byte identical: 9 materialized events and an 864-point UTC-offset sweep (1995–2030, including DST boundaries) match the unpatched library exactly.

## Tests

- `test_get_observance_cache_avoids_repeated_expansion` — deterministic guard (mirrors the `as_rrule` counting test from #592): hammers `get_observance` 4,000 times across two years and asserts `Timezone._observances` is expanded `<= 4` times, plus checks summer/winter offsets are correct.
- `test_benchmark_get_observance_repeated_calls` — `pytest-benchmark` perf regression test for the embedded-VTIMEZONE path.

Full suite passes (`1283 passed`); `ruff`, `ruff format`, `ty`, and `codespell` are clean.

Refs #593. Complements #592.

While I am clearly out of the depth when it comes to the logic. I am happy to invest time & token to refactor and test to your liking.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: tests/test_timezone.py [hunk 1]
```diff
@@ -216,3 +218,99 @@ def test_clear_old_dtstamp(mock_prodid: Generator[None, None, None]) -> None:
...
+    winter = timezone.get_observance(datetime.datetime(2024, 1, 15, 12, 0, 0))
+    assert summer is not None
+    assert summer.observance.tz_offset_to.offset == datetime.timedelta(hours=2)
+    assert winter is not None
+    assert winter.observance.tz_offset_to.offset == datetime.timedelta(hours=1)
+
+    original = Timezone._observances
+    call_count = 0
+
+    def counting_observances(self: Timezone) -> object:
```
Comment (allenporter) [inline tests/test_timezone.py]: Let's not have a test muck with internal details of the class. Given this is a performance fix, lets just rely on the benchmark to measure performance -- and leave tests to measure correctness of the code.

File: ical/timezone.py [hunk 2]
```diff
@@ -249,12 +263,23 @@ def get_observance(self, value: datetime.datetime) -> _ObservanceInfo | None:
         """Return the specified observance for the specified date."""
         if value.tzinfo is not None:
             raise ValueError("Start time must be in local time format")
-        last_observance_info: _ObservanceInfo | None = None
-        for dt_start, observance_info in self._observances():
-            if dt_start > value:
-                return last_observance_info
-            last_observance_info = observance_info
-        return last_observance_info
+        cache = self._observance_cache
```
Comment (allenporter) [inline ical/timezone.py]: My understanding is that this is a class level variable, so it's shared across instances of this class, is that right? So given that, this solution is not safe/correct, right? It's setting a global observance for any date time independent of the actual timezone.

Generally, I am familiar with the performance problem here given the office calendar custom timezones start so early, so supportive on collaborating on a fix.

**Review by allenporter — CHANGES_REQUESTED** [review]
Appreciate the diagnosis and proposed solution, though i have some comments about it below. i can get more prescriptive on solutions, but wanted to know how much you and claude want to lean in...

File: tests/test_timezone.py (hunk 1 above)
Comment (FuNK3Y) [reply tests/test_timezone.py]: Agreed — dropped that test in 8064c4a. It now relies on the benchmark for the performance dimension and a plain correctness test (`test_customized_timezone_observance_offsets`) that just checks resolved offsets, no patching of internals.

File: ical/timezone.py (hunk 2 above)
Comment (FuNK3Y) [reply ical/timezone.py]: Good thing to double-check. It reads like a class attribute, but `PrivateAttr(default_factory=list)` is pydantic's idiom for **per-instance** mutable state: the factory runs per object and the value lives in `self.__pydantic_private__`, so each `Timezone` gets its own cache — nothing is shared across instances or timezones.

Verified directly with two different custom zones (EU `+1/+2` and US-Eastern `-5/-4`):

```
eu cache id: …551808   us cache id: …058048   same object? False
eu summer: 2:00:00   us summer: -4h     eu winter: 1:00:00   us winter: -5h
```

I added a comment in 8064c4a calling out the per-instance semantics so it's not a trap for the next reader. Happy to switch to a different caching approach if you'd prefer one — and yep, glad to lean in on this.

**Comment by FuNK3Y** [comment]
So, I did another pass with Claude with your early comments. Be as prescriptive as you feel necessary. While I'd love to have this fixed the maintainability of the overall library is what matter most !

File: ical/timezone.py (hunk 2 above)
Comment (allenporter) [reply ical/timezone.py]: OK thanks for clarifying, sounds good.

I'd like to adjust the abstraction and caching approach before merging:

1. **Separate concerns in `iter.py`:** Let's extract the caching logic out of `timezone.py` into a generic `CachedTransitionTimeline` class at the end of `ical/iter.py` (similar to the other timeline helpers there). It wraps any sorted transition iterable of `(onset, item)` pairs and exposes an `active_at(value)` method. It lazily materializes and caches transitions only up to the queried datetime.
2. **Simplified timezone state:** With that helper, `Timezone` only needs a single private attribute to manage its cache:
   ```python
   _observance_timeline: Optional[CachedTransitionTimeline["_ObservanceInfo"]] = (
       PrivateAttr(default=None)
   )
   ```
3. **Extend the cache, don't reset it (and keep it deepcopy-safe):** Now that the caching is isolated in `CachedTransitionTimeline`, we can also do the optimization to optimize the cache extension:
   * Instead of clearing the cache list on a miss, we can lazily append new transitions.
   * You may need to also support `__deepcopy__` though (since `dateutil.rrule` generators contain non-copyable locks))


WDYT?

**Review by allenporter — CHANGES_REQUESTED** [review]

File: ical/timezone.py (hunk 2 above)
Comment (FuNK3Y) [reply ical/timezone.py]: Done in 64b8354 — implemented all three:

1. **`CachedTransitionTimeline` in `iter.py`** — generic over `(onset, item)`, exposes `active_at(value)`, materializes transitions lazily and caches up to the highest value queried.
2. **`Timezone`** now holds a single private attr:
   ```python
   _observance_timeline: Optional[CachedTransitionTimeline["_ObservanceInfo"]] = (
       PrivateAttr(default=None)
   )
   ```
   `get_observance` just lazily constructs it from `self._observances` and delegates to `active_at`.
3. **Extend, don't reset** — the cache is append-only and the iterator is never restarted. For deepcopy-safety I added `CachedTransitionTimeline.__deepcopy__`: since `dateutil` iterators carry non-copyable state, the copy doesn't clone the live iterator/cache — it shares the (pure) transition factory and rebuilds lazily. Verified `copy.deepcopy` of a warmed `Timezone` and the real `IcsTimezoneInfo.__deepcopy__` path both still resolve correct offsets.

Added unit tests for the helper (lookup ordering, single lazy consumption, deepcopy independence) in `tests/test_iter.py`. Full suite green (1286), ruff/ty/codespell clean. `materialize_timeline` on the original Office 365 calendar stays at ~0.005s.

Thanks for the clear direction — happy to tweak the abstraction further if you'd like.

**Comment by FuNK3Y** [comment]
Here it is - taking your comments into account.

File: ical/iter.py [hunk 3]
```diff
@@ -437,3 +439,54 @@ def as_rrule(
...
+    ``active_at`` returns the item of the latest transition whose onset is at or
+    before the queried value. Transitions are materialized lazily and appended
+    to a cache only as far as the highest value queried so far, so an unbounded
+    recurrence (such as a timezone observance rule starting in year 1601) is
+    expanded once instead of on every lookup.
+    """
+
+    def __init__(
+        self,
+        transitions: Callable[[], Iterable[tuple[Any, T]]],
```
Comment (allenporter) [inline ical/iter.py]: I think this can accept an Iterable here rather than a callable? (and update the caller). I consider an Iterable already to have the similar properties (e.g. its kind of like a Callable to get back an iterator i think)

File: ical/iter.py [hunk 4]
```diff
@@ -437,3 +439,54 @@ def as_rrule(
...
+    ) -> None:
+        """Initialize CachedTransitionTimeline with a transition source factory."""
+        self._transitions = transitions
+        self._iter: Iterator[tuple[Any, T]] | None = None
+        self._cache: list[tuple[Any, T]] = []
+        self._exhausted = False
+
+    def active_at(self, value: Any) -> T | None:
```
Comment (allenporter) [inline ical/iter.py]: Can we do better than `Any` here and in the constructor? If needed could limit to date/time types (i assume making these sortable items isn't worth it)

File: ical/iter.py [hunk 5]
```diff
@@ -437,3 +439,54 @@ def as_rrule(
         rdate,
         exdate,
     )
+
+
+class CachedTransitionTimeline(Generic[T]):
```
Comment (allenporter) [inline ical/iter.py:444]: I know before i said not to test the details of the internals but perhaps now that its a separate class (and i asked you to make it more complex) a couple extra corner case tests could be useful (still without digging into internals, but just verifying correctness of the cache in different request order scenarios)

**Review by allenporter — CHANGES_REQUESTED** [review]
This is great, much appreciated. One last round and i think we're good to merge.

File: ical/iter.py (hunk 3 above)
Comment (FuNK3Y) [reply ical/iter.py]: Done in f77d73d — now takes `Iterable[tuple[onset, item]]` and the caller passes `self._observances()` directly.

One subtlety worth recording: the source must be **re-iterable**, because `__deepcopy__` can't copy it. Every `dateutil` rrule carries a `threading.Lock`, so deep-copying the transition source fails with `cannot pickle '_thread.lock' object`. So the copy *shares* the source and takes its own fresh `iter()` — which is safe precisely because `MergedIterable` is re-iterable (I verified independent/concurrent traversals stay consistent). I noted that contract in the docstring.

File: ical/iter.py (hunk 4 above)
Comment (FuNK3Y) [reply ical/iter.py]: Done — typed the onset as `datetime.datetime | datetime.date` (matching `_observances()`'s element type) in the constructor, `active_at`, the cache and the iterator. Left the item as the existing `Generic[T]`. Agreed that making these full `SortableItem`s isn't worth it here.

File: ical/iter.py (hunk 5 above)
Comment (FuNK3Y) [reply ical/iter.py:444]: Added in f77d73d — a parametrized `test_cached_transition_timeline_request_order` covering out-of-order, descending, ascending, repeated and interleaved-extremes request sequences, plus an onset-equality assertion in `test_cached_transition_timeline_active_at`. All via the public `active_at` (no internals). I also kept a lazy-consumption test that asserts the source is iterated once and only pulled far enough to answer the query, and a deepcopy-independence test — both through public behavior. Full suite 1291 green.

**Comment by FuNK3Y** [comment]
Changes implemented.

File: ical/iter.py (hunk 3 above)
Comment (allenporter) [reply ical/iter.py]: > One subtlety worth recording: the source must be **re-iterable**, because `__deepcopy__` can't copy it. Every `dateutil` rrule carries a `threading.Lock`, so deep-copying the transition source fails with `cannot pickle '_thread.lock' object`. So the copy _shares_ the source and takes its own fresh `iter()` — which is safe precisely because `MergedIterable` is re-iterable (I verified independent/concurrent traversals stay consistent). I noted that contract in the docstring.

Yeah i believe here we understand that Iterable must always be re-iterable -- you get a new iterator every time. I had the impression that was the whole point of Iterable vs Iterator. Seems fine to note in the docstring, thank you.

**Review by allenporter — APPROVED** [review]
@FuNK3Y thank you for this awesome performance improvement. I didn't have the energy to support this use case, but i know its pretty popular, so thank you so much for this contribution.

I can release a new version of `ical` if you want to do the update to home assistant?

**Comment by FuNK3Y** [comment]
> @FuNK3Y thank you for this awesome performance improvement. I didn't have the energy to support this use case, but i know its pretty popular, so thank you so much for this contribution.
> 
> I can release a new version of `ical` if you want to do the update to home assistant?

Thank you @allenporter for the responsiveness & guidance. That was fun.

Yes, a release would be awesome. I'll try to squeeze it in the next release of Home Assistant.
